kore/src/Kore/Internal/SideCondition/SideCondition.hs

```diff
@@ -11,43 +11,81 @@ module Kore.Internal.SideCondition.SideCondition
 import Prelude.Kore
 
 import Control.DeepSeq
-    ( NFData
+    ( NFData (..)
     )
 import Data.Hashable
-    ( Hashable
+    ( Hashable (..)
     , Hashed
     , hashed
     )
 import Data.Text
     ( Text
     )
-import qualified Generics.SOP as SOP
-    ( Generic
-    , HasDatatypeInfo
-    )
-import qualified GHC.Generics as GHC
 
+import Data.Type.Equality
+    ( (:~:) (..)
+    , testEquality
+    )
 import Kore.Debug
-    ( Debug
+    ( Debug (..)
     , Diff (..)
     )
+import Type.Reflection
+    ( SomeTypeRep (..)
+    , TypeRep
+    , Typeable
+    , typeRep
+    )
+
+data Representation where
+    Representation :: Ord a => !(TypeRep a) -> !(Hashed a) -> Representation
+
+instance Eq Representation where
+    (==) (Representation typeRep1 hashed1) (Representation typeRep2 hashed2) =
+        case testEquality typeRep1 typeRep2 of
+            Nothing -> False
+            Just Refl -> hashed1 == hashed2
+    {-# INLINE (==) #-}
+
+instance Ord Representation where
+    compare
+        (Representation typeRep1 hashed1)
+        (Representation typeRep2 hashed2)
+      =
+        case testEquality typeRep1 typeRep2 of
+            Nothing -> compare (SomeTypeRep typeRep1) (SomeTypeRep typeRep2)
+            Just Refl -> compare hashed1 hashed2
+    {-# INLINE compare #-}
+
+instance Show Representation where
+    showsPrec prec (Representation typeRep1 _) =
+        showParen (prec >= 10)
+        $ showString "Representation " . shows typeRep1 . showString " _"
```

Comment:
That's exactly it: I'm not sure it would be useful to `show` the condition, and it would dominate the output to the point that I'm afraid it wouldn't be useful.